conductr_cli/sandbox_run.py

```
@@ -42,15 +42,14 @@ def run(args):
     feature_results = []
     print_proxy_output = False
     if is_conductr_started:
-        if not is_conductr_v1:
-            print_proxy_output = sandbox_proxy.start_proxy(proxy_bind_addr=run_result.core_addrs[0],
-                                                           bundle_http_port=args.bundle_http_port,
-                                                           proxy_ports=sorted(args.ports))
-
         for feature in features:
+            feature.configure(args, run_result)
             result = feature.start()
             feature_results += result
 
+            if isinstance(feature, sandbox_features.ProxyingFeature):
```

This I'm a bit iffy about, because Proxying is a "special" feature.

---

Yeah, if feels wrong to do a special test. I thought that features had some facility to contribute to the output. Perhaps look at the monitoring feature for inspiration.

---

I'd suggest introduce a property `self.print_start_output = True` for the `ProxyingFeature`, and `self.print_start_output = False` for everything else. This way you can just check if `feature.print_start_output` is `True`, and print the output accordingly. Wdyt?

---

I think your suggestion to add a property to the feature is right but I'm thinking we can generalize it a bit.

As an idea, maybe features also have a `provides` property (e.g. the `proxy` and `log` class) such that only a single feature can be enabled at a time that has that `provides` entry. Then, this check simply becomes, were any features with `provides` `proxy` started? If so, we alter the service URL displayed. In the future, this would allow multiple feature implementations of the proxying concept and `sandbox_run_jvm.py` wouldn't care which was actually enabled.

Similarly, this would generalize the eslite/es logic. We could throw an error if two features were specified that both provided logging.

Just for reference, this boolean (`print_proxy_output`) is used as the following:

```
               for feature_result in feature_results:
                    if is_proxy_started:
                        uri = '{}:{}'.format(run_result.host, feature_result.port)
                    else:
                        uri = '{}:{}/services/{}'.format(run_result.host, DEFAULT_SERVICE_LOCATOR_PORT,
```

---

I went ahead with my idea here. Features now have a `provides` mechanism to provide any arbitrary string. No two enabled features can provide that same value -- this allows us to ensure only one logging feature is enabled, or in the future only one proxying feature. Let me know what you guys think of this!